### PR TITLE
Initial Tor support

### PR DESCRIPTION
**Pull request by rorp**
I copypasted the modules I wrote for Eclair (https://github.com/ACINQ/eclair/tree/master/eclair-core/src/main/scala/fr/acinq/eclair/tor) with some minimal changes.

### CODE REVIEW

File: tor-test/src/test/scala/org/bitcoins/tor/TorProtocolHandlerSpec.scala [hunk 1]
```diff
@@ -0,0 +1,353 @@
+/*
+ * Copyright 2019 ACINQ SAS
```
Comment (benthecarman) [inline tor-test/src/test/scala/org/bitcoins/tor/TorProtocolHandlerSpec.scala]: Can change

File: tor-test/src/test/scala/org/bitcoins/tor/TorProtocolHandlerSpec.scala [hunk 2]
```diff
@@ -0,0 +1,353 @@
...
+    val promiseOnionAddress = Promise[InetSocketAddress]()
+
+    Files.write(CookieFilePath, CryptoUtil.randomBytes(32).toArray)
+
+    val protocolHandler = TestActorRef(
+      props(
+        version = OnionServiceVersion("v2"),
+        authentication = SafeCookie(ClientNonce),
+        privateKeyPath = PkFilePath,
+        virtualPort = 9999,
```
Comment (benthecarman) [inline tor-test/src/test/scala/org/bitcoins/tor/TorProtocolHandlerSpec.scala:213]: Should use a different port as this is what we use for the rpc server.

File: tor-test/src/test/scala/org/bitcoins/tor/TorProtocolHandlerSpec.scala [hunk 3]
```diff
@@ -0,0 +1,353 @@
...
+    val promiseOnionAddress = Promise[InetSocketAddress]()
+
+    Files.write(CookieFilePath, AuthCookie.toArray)
+
+    val protocolHandler = TestActorRef(
+      props(
+        version = OnionServiceVersion("v2"),
+        authentication = SafeCookie(ClientNonce),
+        privateKeyPath = PkFilePath,
+        virtualPort = 9999,
```
Comment (benthecarman) [inline tor-test/src/test/scala/org/bitcoins/tor/TorProtocolHandlerSpec.scala:248]: Ditto

File: tor-test/src/test/scala/org/bitcoins/tor/TorProtocolHandlerSpec.scala [hunk 4]
```diff
@@ -0,0 +1,353 @@
...
```
Comment (benthecarman) [inline tor-test/src/test/scala/org/bitcoins/tor/TorProtocolHandlerSpec.scala:290]: Ditto

File: tor/src/main/scala/org/bitcoins/tor/Controller.scala [hunk 5]
```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright 2019 ACINQ SAS
```
Comment (benthecarman) [inline tor/src/main/scala/org/bitcoins/tor/Controller.scala]: Can change

File: tor-test/src/test/scala/org/bitcoins/tor/TorProtocolHandlerSpec.scala (hunk 2 above)
Comment (benthecarman) [reply tor-test/src/test/scala/org/bitcoins/tor/TorProtocolHandlerSpec.scala:213]: Can use TestUtil.randomPort

File: tor/src/main/scala/org/bitcoins/tor/Socks5Connection.scala [hunk 6]
```diff
@@ -0,0 +1,266 @@
+/*
+ * Copyright 2019 ACINQ SAS
```
Comment (benthecarman) [inline tor/src/main/scala/org/bitcoins/tor/Socks5Connection.scala]: Can change

File: tor/src/main/scala/org/bitcoins/tor/Socks5Connection.scala [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+    address: InetSocketAddress,
+    credentials_opt: Option[Credentials],
+    randomizeCredentials: Boolean,
+    useForIPv4: Boolean,
+    useForIPv6: Boolean,
+    useForTor: Boolean)
+
+object Socks5ProxyParams {
+
+//  def proxyAddress(
```
Comment (benthecarman) [inline tor/src/main/scala/org/bitcoins/tor/Socks5Connection.scala]: Can remove or uncomment

File: tor/src/main/scala/org/bitcoins/tor/TorProtocolHandler.scala [hunk 8]
```diff
@@ -0,0 +1,365 @@
+/*
+ * Copyright 2019 ACINQ SAS
```
Comment (benthecarman) [inline tor/src/main/scala/org/bitcoins/tor/TorProtocolHandler.scala]: Can change

**Review by benthecarman — COMMENTED** [review]
Quick review for nits

**Review by Christewart — APPROVED** [review]
I'm good with merging this as a starter, but there is a lot of refactoring / integration we will want to do in the future. For instance we probably should add an sbt task called `downloadTor` or something like that to fetch the tor daemon? My understanding is that is required.

**Review by benthecarman — APPROVED** [review]
Agreed, good for now

**Pull request merged**